From b448e73099d38f4a8ac2cdf447b6432ce7dafe56 Mon Sep 17 00:00:00 2001
From: Vladimir Oltean <olteanv@gmail.com>
Date: Mon, 28 May 2018 16:34:10 +0300
Subject: [PATCH 39/41] Make volume steps and defaults adjustable for all audio
 streams

* AOSP permits adjustment of the following properties:

     Stream       |       Number of steps       |       Default volume
------------------+-----------------------------+----------------------------
STREAM_VOICE_CALL | ro.config.vc_call_vol_steps |             -
   STREAM_MUSIC   |  ro.config.media_vol_steps  | ro.config.media_vol_default
   STREAM_ALARM   |              -              |             -
   STREAM_RING    |              -              |             -

* This patch effectively completes the table of supported adjustments by
  introducing the following new properties:
    - ro.config.alarm_vol_steps, ro.config.alarm_vol_default
    - ro.config.ring_vol_steps,  ro.config.ring_vol_default
    - ro.config.vc_call_vol_default
* This allows a device that used one of the 3 existing properties to
  have a uniform level of control over all streams in the volume widget.

Change-Id: I8c9a6f1b34212f575672d3f52f4d7c71ecd31936
Signed-off-by: Vladimir Oltean <olteanv@gmail.com>
---
 .../android/server/audio/AudioService.java    | 75 +++++++++++++------
 1 file changed, 52 insertions(+), 23 deletions(-)

diff --git a/services/core/java/com/android/server/audio/AudioService.java b/services/core/java/com/android/server/audio/AudioService.java
index f7b139ec0a0..762b7b7b0ce 100644
--- a/services/core/java/com/android/server/audio/AudioService.java
+++ b/services/core/java/com/android/server/audio/AudioService.java
@@ -687,30 +687,59 @@ public class AudioService extends IAudioService.Stub
         mHasVibrator = vibrator == null ? false : vibrator.hasVibrator();
 
         // Initialize volume
-        int maxCallVolume = SystemProperties.getInt("ro.config.vc_call_vol_steps", -1);
-        if (maxCallVolume != -1) {
-            MAX_STREAM_VOLUME[AudioSystem.STREAM_VOICE_CALL] = maxCallVolume;
-            AudioSystem.DEFAULT_STREAM_VOLUME[AudioSystem.STREAM_VOICE_CALL] =
-                    (maxCallVolume * 3) / 4;
-        }
-
-        int maxMusicVolume = SystemProperties.getInt("ro.config.media_vol_steps", -1);
-        if (maxMusicVolume != -1) {
-            MAX_STREAM_VOLUME[AudioSystem.STREAM_MUSIC] = maxMusicVolume;
-        }
-
-        int defaultMusicVolume = SystemProperties.getInt("ro.config.media_vol_default", -1);
-        if (defaultMusicVolume != -1 &&
-                defaultMusicVolume <= MAX_STREAM_VOLUME[AudioSystem.STREAM_MUSIC]) {
-            AudioSystem.DEFAULT_STREAM_VOLUME[AudioSystem.STREAM_MUSIC] = defaultMusicVolume;
-        } else {
-            if (isPlatformTelevision()) {
-                AudioSystem.DEFAULT_STREAM_VOLUME[AudioSystem.STREAM_MUSIC] =
-                        MAX_STREAM_VOLUME[AudioSystem.STREAM_MUSIC] / 4;
-            } else {
-                AudioSystem.DEFAULT_STREAM_VOLUME[AudioSystem.STREAM_MUSIC] =
-                        MAX_STREAM_VOLUME[AudioSystem.STREAM_MUSIC] / 3;
+        String volumeStepsProperties[] = {
+            "ro.config.vc_call_vol_steps",
+            "ro.config.media_vol_steps",
+            "ro.config.alarm_vol_steps",
+            "ro.config.ring_vol_steps",
+        };
+        String volumeDefaultsProperties[] = {
+            "ro.config.vc_call_vol_default",
+            "ro.config.media_vol_default",
+            "ro.config.alarm_vol_default",
+            "ro.config.ring_vol_default",
+        };
+        int streams[] = {
+            AudioSystem.STREAM_VOICE_CALL,
+            AudioSystem.STREAM_MUSIC,
+            AudioSystem.STREAM_ALARM,
+            AudioSystem.STREAM_RING,
+        };
+        int defaultStreamVolumePercents[][] = {
+            {       // Android TV
+                75, // STREAM_VOICE_CALL
+                25, // STREAM_MUSIC
+                25, // STREAM_ALARM
+                25, // STREAM_RING
+            }, {    // Regular Android
+                75, // STREAM_VOICE_CALL
+                33, // STREAM_MUSIC
+                33, // STREAM_ALARM
+                33, // STREAM_RING
+            },
+        };
+        for (int i = 0; i < volumeStepsProperties.length; i++) {
+            int maxVolume = SystemProperties.getInt(volumeStepsProperties[i], -1);
+            if (maxVolume == -1) {
+                // ro.config.*_vol_steps not specified, apply defaults.
+                maxVolume = MAX_STREAM_VOLUME[streams[i]];
+            }
+            MAX_STREAM_VOLUME[streams[i]] = maxVolume;
+        }
+        for (int i = 0; i < volumeDefaultsProperties.length; i++) {
+            int volume = SystemProperties.getInt(volumeDefaultsProperties[i], -1);
+
+            if (volume == -1 || (volume > MAX_STREAM_VOLUME[streams[i]])) {
+                // ro.config.*_default not specified, apply defaults.
+                int percent;
+                if (isPlatformTelevision()) {
+                    percent = defaultStreamVolumePercents[0][i];
+                } else {
+                    percent = defaultStreamVolumePercents[1][i];
+                }
+                volume = MAX_STREAM_VOLUME[streams[i]] * percent / 100;
             }
+            AudioSystem.DEFAULT_STREAM_VOLUME[streams[i]] = volume;
         }
 
         sSoundEffectVolumeDb = context.getResources().getInteger(
-- 
2.17.1

